cohortextractor/cohortextractor.py

```diff
@@ -117,23 +119,92 @@ def generate_cohort(output_dir, expectations_population, selected_study_name=Non
                 break
     for study_name, suffix in study_definitions:
         print(f"Generating cohort for {study_name}...")
-        _generate_cohort(output_dir, study_name, suffix, expectations_population)
+        _generate_cohort(
+            output_dir,
+            study_name,
+            suffix,
+            expectations_population,
+            index_date_range=index_date_range,
+        )
 
 
-def _generate_cohort(output_dir, study_name, suffix, expectations_population):
+def _generate_cohort(
+    output_dir, study_name, suffix, expectations_population, index_date_range=None
+):
     print("Running. Please wait...")
     study = load_study_definition(study_name)
 
     with_sqlcmd = shutil.which("sqlcmd") is not None
     os.makedirs(output_dir, exist_ok=True)
-    study.to_csv(
-        f"{output_dir}/input{suffix}.csv",
-        expectations_population=expectations_population,
-        with_sqlcmd=with_sqlcmd,
-    )
-    print(
-        f"Successfully created cohort and covariates at {output_dir}/input{suffix}.csv"
-    )
+    for index_date in _generate_date_range(index_date_range):
+        if index_date is not None:
+            study.set_index_date(index_date)
+            date_suffix = f"_{index_date}"
+        else:
+            date_suffix = ""
+        output_file = f"{output_dir}/input{suffix}{date_suffix}.csv"
+        study.to_csv(
+            output_file,
+            expectations_population=expectations_population,
+            with_sqlcmd=with_sqlcmd,
+        )
+        print(f"Successfully created cohort and covariates at {output_file}")
+
+
+def _generate_date_range(date_range_str):
+    # Bail out with an "empty" range: this means we don't need separate
+    # codepaths to handle the range, single date, and no date supplied cases
+    if not date_range_str:
+        return [None]
+    start, end, period = _parse_date_range(date_range_str)
+    if end < start:
+        raise ValueError(
+            f"Invalid date range '{date_range_str}': end cannot be earlier than start"
+        )
+    dates = []
+    while start <= end:
+        dates.append(start.isoformat())
+        start = _increment_date(start, period)
+    return dates
+
+
+def _parse_date_range(date_range_str):
+    period = "month"
+    if " to " in date_range_str:
+        start, end = date_range_str.split(" to ", 1)
+        if " by " in end:
+            end, period = end.split(" by ", 1)
+    else:
+        start = end = date_range_str
+    try:
+        start = _parse_date(start)
+        end = _parse_date(end)
+    except ValueError:
+        raise ValueError(
+            f"Invalid date range '{date_range_str}': Dates must be in YYYY-MM-DD "
+            f"format or 'today' and ranges must be in the form "
+            f"'DATE to DATE by (week|month)'"
+        )
```

Comment:
nit: slightly better to validate `week|month` here rather than when it's attempted to be used in `_increment_date`?